src/HealthChecks/HealthChecks/src/HealthCheckPublisherHostedService.cs

```diff
@@ -131,13 +178,33 @@ internal async Task RunAsync()
         }
     }
 
-    private async Task RunAsyncCore(CancellationToken cancellationToken)
+    private async Task RunAsyncCore((TimeSpan Delay, TimeSpan Period) timerOptions, CancellationToken cancellationToken)
     {
         // Forcibly yield - we want to unblock the timer thread.
         await Task.Yield();
 
+        // Concatenate predicates - we only run HCs at the set delay, period and timeout, and that are enabled
+        var withOptionsPredicate = (HealthCheckRegistration r) =>
+        {
+            // Check whether the current timer options correspond to the ones of the HC registration
+            var rOptions = GetTimerOptionsOrDefault(r);
+            var hasOptions = rOptions == timerOptions;
+            if (!hasOptions)
+            {
+                return false;
+            }
+
+            if (_healthCheckPublisherOptions?.Value.Predicate == null)
+            {
+                return true;
+            }
+
+            // Else check the user-applied predicates
+            return _healthCheckPublisherOptions.Value.Predicate(r);
```

Comment:
Nit. I think code golfing actually improves readability here.

```suggestion
            // First check whether the current timer options correspond to the current registration,
            // and then check the user-defined predicate if any.
            return GetTimerOptions(r) == timerOptions &&
                (_healthCheckPublisherOptions?.Value.Predicate(r) ?? true);
```

Reply:
Changed to:

```csharp
// First check whether the current timer options correspond to the current registration,
// and then check the user-defined predicate if any.
return (GetTimerOptions(r) == timerOptions) && (_healthCheckPublisherOptions?.Value.Predicate ?? (_ => true))(r);